### ISSUE TITLE
MVCC parallelism 

### ISSUE THREAD

**Issue by can-keklik** [erikgrinaker/toydb#60]
Hey @erikgrinaker,

I read the architecture guide but maybe I missed this one somehow. I was wondering how MVCC storage behaves in terms of parallelism. I see that there is a Mutex here. So how does that affect execution exactly? Is there really only one transaction running at a time? Like, yeah, this means transactions are concurrent, but still not parallel as I understood.

Sorry if I'm missing something fundamental.

```rust
pub struct MVCC<E: Engine> {
    /// The underlying KV engine. It is protected by a mutex so it can be shared between txns.
    engine: Arc<Mutex<E>>,
}
```
Thanks in advance.

**Comment by erikgrinaker** [comment]
That's right, only a single operation can be performed on the MVCC engine at a time. There's a reason this is called a toy. :) 

This follows from the fact that the storage engine isn't thread-safe. But parallelism isn't really needed either, since Raft will serialize all state operations anyway. In principle, it would be possible to do concurrent reads, but it isn't worth the effort.

**Comment by can-keklik** [comment]
Oh, I see. So can we summarize it like:

Parallel writes are not possible and
Parallel reads can be achieved via replication?

**Comment by erikgrinaker** [comment]
Not quite. There's no way currently to read from Raft followers. But the point was that, in principle, it is possible to do concurrent reads from a Raft state machine as long as the storage engine and query execution engine supports it. toyDB currently doesn't support this though, all reads are processed serially.

**Comment by can-keklik** [comment]
My bad. Let me rephrase what I understood :) 

- So replication is just for fault tolerance I suppose. 
- Even if Raft allowed parallel execution, reads would still be only concurrent, since MVCC engine is not thread-safe.

**Comment by erikgrinaker** [comment]
Yep, correct. There is no horizontal scalability in toyDB -- that would require linearizable follower reads and/or running multiple Raft groups with a distributed transaction protocol (which is what grown-up distributed databases do).

**Comment by can-keklik** [comment]
Thanks, I'm a huge fan of this project.

**Comment by erikgrinaker** [comment]
Thanks! Will close this out then, but let me know if you have any further questions.
